engine/storage/volume/src/main/java/org/apache/cloudstack/storage/volume/VolumeServiceImpl.java

```diff
@@ -522,6 +522,21 @@ public Void deleteVolumeCallback(AsyncCallbackDispatcher<VolumeServiceImpl, Comm
         return null;
     }
 
+    /**
+     * Deletes the snapshot from primary storage if the only storage associated with the snapshot is of the Primary role; else, just removes the primary record on the DB.
+     * */
+    protected void deleteKvmSnapshotOnPrimary(SnapshotDataStoreVO snapshotDataStoreVO) {
+        List<SnapshotDataStoreVO> snapshotDataStoreVOList = _snapshotStoreDao.findBySnapshotId(snapshotDataStoreVO.getSnapshotId());
+        for (SnapshotDataStoreVO snapshotStore : snapshotDataStoreVOList) {
+            if (DataStoreRole.Image.equals(snapshotStore.getRole())) {
+                _snapshotStoreDao.remove(snapshotDataStoreVO.getId());
+                return;
+            }
+        }
+
+        snapshotApiService.deleteSnapshot(snapshotDataStoreVO.getSnapshotId(), null);
```

Comment:
yes, I know about the garbage left but I mean do we want to delete them

Reply:
@slavkap I believe that keeping snapshots on primary storage is meant mainly as a fast(er) way to rewind a volume. When the volume is removed, the snapshots on primary kind of lose their purpose, so they are removed. However, this could be debated further in another issue or the ML. In any case, changing this behavior is our of scope for this PR, I only added the file deletion as a bug fix. If need be, I can extract it into another PR.